### ISSUE TITLE
'default' and 'required' 

### ISSUE THREAD

**Issue by ahollinshead** [apicollective/apibuilder#454]
I assumed that when a field had a default value that it wouldn't have to be passed to the server, that the server would use the default value in cases when it wasn't passed.

When I upload an api.json with a field with a default

```
        { "name": "user_partition", "type": "integer", "default": 1 },
```

it is showing in the UI as being required 'no', however when I generate play code, the json parser does not allow the field to be nullable. So the service gives an error if you don't include that field when you post an instance of the model without that field, even though it works fine for non-required fields that don't have defaults.

If I explicitly add "required": "false" and try to upload it tells me that a field with a default must be required (which conflicts with the UI).

It's unclear what use the 'default' property is serving. Is support for it only used in generated clients? (I wasn't using a generated client). If so, the name should maybe change to represent this. And the apidoc UI definitely shouldn't be saying required is 'no' in that case. It feels like it would be a lot more useful if default actually meant that you didn't have the pass it to the server.

**Comment by mbryzek** [comment]
We had a lot of discussion about this internally in the early days - and in fact the early implementation was what you described.

Where we ended up was deciding that it made more sense for:

  - required to indicate whether or not a value must always be present (no confusion here)
  - default values would be interpreted as being required - if there is a default, that value must always be present

@haywood and I had a lot of debate on this in the early days of iris hub :) but ultimately this approach resolved a number of inconsistencies (hoping he may remember some of the specific details).

In practice, this approach makes sure that defaults are applied consistently; what we have done in situtations like you describe is:

Either:
  - make the field required: false w/ no default int the definition of the form body sent to server and just set the default server side
  - specify a default (and have it sent everywhere)

**Comment by gheine** [comment]
Ok, I think these are two separate issues:
1) The UI shows fields with default values as not required (Required? column: No) 
2) Should we support default values for non-required fields? Ie. not require the field to be present on the wire (in JSON) but prepresent them as non-optional types in the generated model code, with the de-serializer filling in the default value provided in the API if the field is not present on the wire.

**Comment by marioduarte** [comment]
The way I look at this there should really 2 options:
* `required = true` - This field needs to be sent on the http request, and if it is missing, it is an invalid request. It does not make sense to have a default value in this case.
* `required = false` - This field does NOT need to be sent on the http request. If the value is missing the server may have a default value for this field. A default value as I see it only makes sense as a way to handle on the server side the absence of the field. This means that the `default` property would not be used in any way in the code generation of the data models of the clients.

The way I see it Apibuilder should provide the definition of HTTP rest endpoints, regardless of whether one uses the generated clients or not. Having a field marked as Not required (with a default value) should mean that the field does not need to be sent to the server instead of meaning that the generated clients will quietly sent the default value to the server.

**Comment by gheine** [comment]
Circling back on this again :) I've created #686 to address the UI issue, where an on-the-wire required field is listed as not required.

As far as applying default values with required vs non-required fields, I agree with @marioduarte, with the following minor difference:

* `required = true` - This field needs to be sent over the wire, and if it is missing, it is an invalid request. If a default value is defined, it must be applied on the client side.
* `required = false` - This field does NOT need to be sent on the wire. If the value is missing the server may apply a default value for this field. 

With that, we should add support for non-required fields with default values. These fields would be optional client-side (with no default value substitution), optional on the wire, and the server would use the default value if the field is missing.

**Comment by mbryzek** [comment]
Related PR: https://github.com/apicollective/apibuilder-generator/pull/378

**Comment by gheine** [comment]
Fixed in #378
